earth_enterprise/src/scons/getversion.py

```diff
@@ -42,16 +43,125 @@ def GetLongVersion(backupFile, label=''):
 
 
 def _GitGeneratedLongVersion():
-    """Take the raw information parsed by git, and use it to
-       generate an appropriate version string for GEE."""
+    """Calculate the version name and build number into a single build string."""
 
-    raw = _GetCommitRawDescription()
+    versionName, buildNumber = _GitVersionNameAndBuildNumber()
+    return "{0}-{1}".format(versionName, buildNumber)
 
-    # For tagged commits, return the tag itself
-    if _IsCurrentCommitTagged(raw):
-        return _VersionForTaggedHead(raw)
+
+def _GitCommitCount(tagName='HEAD', baseRef=''):
+    """calculate git commit counts"""
+    repo = _GetRepository()
+    if not baseRef:
+        return len(list(repo.iter_commits(tagName)))
+    else:
+        return len(list(repo.iter_commits(baseRef + '..' + tagName)))
+
+
+def _GitVersionNameAndBuildNumber():
+    """Get the version name and build number based on state of git
+    Use a tag only if HEAD is directly pointing to it and it is a
+    release build tag (see _GetCommitRawDescription for details)
+    otherwise use the branch name"""
+
+    # For tagged commits use the tag
+    rawDescription = _GetCommitRawDescription()
+    if _IsCurrentCommitTagged(rawDescription):
+        # Extract version name and build number
+        # from the tag (should be a release build tag)
+        splitTag = rawDescription.split('-')
+        return splitTag[0], splitTag[1]
     else:
-        return _VersionFromTagHistory(raw)
+        # Use branch name if we are not a detached HEAD
+        branchName = _GitBranchName()
+        if not branchName:
+            # we are a detached head not on a tag so just treat the
+            # raw describe like a topic branch name
+            return rawDescription, str(_GitCommitCount())
+        else:
+            # Get the version name from the branch name
+            if _IsReleaseBranch(branchName):
+                versionName = _GetReleaseVersionName(branchName)
+                return versionName, '{0}.{1}'.format(_GitBranchedBuildNumber(versionName), _GitCommitCount('HEAD', versionName))
+            else:
+                return _sanitizeBranchName(branchName),  str(_GitCommitCount())
+
+
+def _GitBranchedBuildNumber(versionName):
+    """Returns what the build number was from the branch point"""
+    prevRelTag = _GitPreviousReleaseTag(versionName)
+    prevBuildNumber = ''
```

Comment:
I think something besides "BuildNumber" would be appropriate here since we're counting tags, not builds. Maybe "prevPatchNumber"

Reply:
Okay.  How about using the name 'commitCount' as that is what we are doing, counting commits.

Reply:
If I understand correctly, it's getting the number from the tag. Does the tag contain the commit count? I thought it was a patch number like we are doing now.

Reply:
Yes the commit count is known by the tag.  With the new way the commit count replaces the patch number with a build number based on commit counts.  It will be similar to how SVN does revision numbers.

Reply:
I think you'd still want to differentiate between `prevCommitCount` and current `commitCount` though

Reply:
I do that.  This call is only concerned with getting the commit count(s) of a point from where the branch started.  The commit count for the release branch will be appended by the caller (see above).  For example the first build of `5.2.5` will have a long version of something like `5.2.5-707.0` the `707` is the commit count of master from the point the release branch was created.  The `0` is the count since the release branch was created.  If someone fixes a bug and a single commit for that fix is merged into the release branch then the long build number would be `5.2.5-707.1`.

Reply:
Cool, thanks for the explanation